discovery/module.go

```diff
@@ -114,51 +120,64 @@ func (m *Module) Add(serviceID string, presentation vc.VerifiablePresentation) e
-	// VPs should not be valid for too short; unnecessary overhead at the server and clients.
-	// Also protects against expiration not being set at all. The factor is somewhat arbitrary.
-	minValidity := float64(definition.PresentationMaxValidity / 4)
-	if expiration.Sub(time.Now()).Seconds() < minValidity {
-		return fmt.Errorf("presentation is not valid for long enough (min %s)", time.Duration(minValidity)*time.Second)
+	if expiration.IsZero() {
+		return errors.New("presentation does not have an expiration")
 	}
 	// VPs should not be valid for too long, as that would prevent the server from pruning them.
 	if int(expiration.Sub(time.Now()).Seconds()) > definition.PresentationMaxValidity {
 		return fmt.Errorf("presentation is valid for too long (max %s)", time.Duration(definition.PresentationMaxValidity)*time.Second)
 	}
-
+	// Check if the presentation already exists
+	credentialSubjectID, err := credential.PresentationSigner(presentation)
+	if err != nil {
+		return err
+	}
+	exists, err := m.store.exists(serviceID, credentialSubjectID.String(), presentation.ID.String())
+	if err != nil {
+		return err
+	}
+	if exists {
+		return ErrPresentationAlreadyExists
+	}
+	// Depending on the presentation type, we need to validate different properties before storing it.
 	if presentation.IsType(retractionPresentationType) {
-		return m.addRetraction(definition.ID, presentation)
+		err = m.validateRetraction(definition.ID, presentation)
 	} else {
-		return m.addPresentation(definition, presentation)
+		err = m.validateRegistration(definition, presentation)
 	}
+	if err != nil {
+		return err
+	}
+	// Check signature of presentation and contained credential(s)
+	_, err = m.vcrInstance.Verifier().VerifyVP(presentation, true, true, nil)
+	if err != nil {
+		return fmt.Errorf("presentation verification failed: %w", err)
+	}
+	return m.store.add(definition.ID, presentation, nil)
 }
 
-func (m *Module) addPresentation(definition Definition, presentation vc.VerifiablePresentation) error {
-	// Must contain credentials
-	if len(presentation.VerifiableCredential) == 0 {
-		return errors.New("presentation must contain at least one credentialRecord")
-	}
+func (m *Module) validateRegistration(definition ServiceDefinition, presentation vc.VerifiablePresentation) error {
 	// VP can't be valid longer than the credentialRecord it contains
 	expiration := presentation.JWT().Expiration()
 	for _, cred := range presentation.VerifiableCredential {
-		exp := cred.JWT().Expiration()
-		if !exp.IsZero() && expiration.After(exp) {
-			return fmt.Errorf("presentation is valid longer than the credentialRecord(s) it contains")
+		if cred.ExpirationDate != nil && expiration.After(*cred.ExpirationDate) {
+			return fmt.Errorf("presentation is valid longer than the credential(s) it contains")
 		}
 	}
-	// VP must fulfill the PEX Presentation Definition
+	// VP must fulfill the PEX Presentation ServiceDefinition
+	// TODO: Use Validate() once PR is merged
```

Comment:
Do you want to wait for that?

Reply:
Probably not. I'll make an issue.

Reply:
We can't use it, since we don't have a submission. So we'll have to use match.